core/src/main/java/io/undertow/server/handlers/accesslog/DefaultAccessLogReceiver.java

```diff
@@ -309,11 +325,59 @@ public void rotate() {
         }
     }
 
+    @SuppressWarnings("static-access")
     @Override
     public void close() throws IOException {
-        closed = true;
-        if (stateUpdater.compareAndSet(this, 0, 1)) {
-            logWriteExecutor.execute(this);
+        this.closed = true;
+        if (this.stateUpdater.compareAndSet(this, 0, 3)) {
+            flushAndTerminate();
+            return;
+        } else {
+            // state[1,2] - scheduled or running, attempt schedule hijack
+            if (this.stateUpdater.compareAndSet(this, 1, 3)) {
+                //this means this thread raced against scheduled run(). run() will exist ASAP
+                //as 1->2 wont be possible, we are at 3 and this.closed
+                flushAndTerminate();
+                return;
+            }
+            // either failed race to 1->3 or we were in 2. We have to wait here sometime.
+            // wait ~1s, if situation does not clear up, try dumping stuff
+            for(int i=0; i<20;i++) {
+                try {
+                    Thread.currentThread().sleep(50);
+                } catch (InterruptedException e) {
+                    // TODO Auto-generated catch block
+                    e.printStackTrace();
+                }
+                if(this.stateUpdater.get(this) == 3) {
```

Comment:
there is a possibility that we exit run without having the status 3. The reason is that run does not work as a single unit, we are not synchronized. So, if we set closed to true too late in the game, we risk being at 2 while sleeping and waking up in a 0 or 1 scenario